From 79299de1c7776ebcbdf8eca1829b1c816e83aeae Mon Sep 17 00:00:00 2001
From: Nishanth Menon <nm@ti.com>
Date: Mon, 28 Feb 2011 22:56:09 +0530
Subject: [PATCH 56/56] sr1.5: dont trust sr layer

SR layer is still in infancy - goofups in terms of NULL pointer passing
can exist. better to check them instead of being burnt.

Signed-off-by: Nishanth Menon <nm@ti.com>
---
 arch/arm/mach-omap2/smartreflex-class1p5.c |   31 +++++++++++++++++++++++++--
 1 files changed, 28 insertions(+), 3 deletions(-)

diff --git a/arch/arm/mach-omap2/smartreflex-class1p5.c b/arch/arm/mach-omap2/smartreflex-class1p5.c
index 6bddce0..6810da9 100644
--- a/arch/arm/mach-omap2/smartreflex-class1p5.c
+++ b/arch/arm/mach-omap2/smartreflex-class1p5.c
@@ -102,8 +102,13 @@ static int sr_class1p5_notify(struct voltagedomain *voltdm, u32 status)
 {
 	struct sr_class1p5_work_data *work_data;
 	int idx = 0;
-	work_data = get_sr1p5_work(voltdm);
 
+	if (IS_ERR_OR_NULL(voltdm)) {
+		pr_err("%s: bad parameters!\n", __func__);
+		return -EINVAL;
+	}
+
+	work_data = get_sr1p5_work(voltdm);
 	if (unlikely(!work_data)) {
 		pr_err("%s:%s no work data!!\n", __func__, voltdm->name);
 		return -EINVAL;
@@ -309,6 +314,12 @@ static int sr_class1p5_enable(struct voltagedomain *voltdm,
 {
 	int r;
 	struct sr_class1p5_work_data *work_data;
+
+	if (IS_ERR_OR_NULL(voltdm) || IS_ERR_OR_NULL(volt_data)) {
+		pr_err("%s: bad parameters!\n", __func__);
+		return -EINVAL;
+	}
+
 	/* if already calibrated, nothing to do here.. */
 	if (volt_data->volt_calibrated)
 		return 0;
@@ -356,6 +367,11 @@ static int sr_class1p5_disable(struct voltagedomain *voltdm,
 {
 	struct sr_class1p5_work_data *work_data;
 
+	if (IS_ERR_OR_NULL(voltdm) || IS_ERR_OR_NULL(volt_data)) {
+		pr_err("%s: bad parameters!\n", __func__);
+		return -EINVAL;
+	}
+
 	work_data = get_sr1p5_work(voltdm);
 	if (work_data->work_active) {
 		/* if volt reset and work is active, we dont allow this */
@@ -387,6 +403,10 @@ static int sr_class1p5_disable(struct voltagedomain *voltdm,
  */
 static int sr_class1p5_configure(struct voltagedomain *voltdm)
 {
+	if (IS_ERR_OR_NULL(voltdm)) {
+		pr_err("%s: bad parameters!\n", __func__);
+		return -EINVAL;
+	}
 	return sr_configure_errgen(voltdm);
 }
 
@@ -448,8 +468,8 @@ static int sr_class1p5_cinit(struct voltagedomain *voltdm,
 	struct sr_class1p5_work_data *work_data;
 	int idx;
 
-	if (!class_priv_data) {
-		pr_err("%s: bad param? no priv data!\n", __func__);
+	if (IS_ERR_OR_NULL(voltdm) || IS_ERR_OR_NULL(class_priv_data)) {
+		pr_err("%s: bad parameters!\n", __func__);
 		return -EINVAL;
 	}
 
@@ -491,6 +511,11 @@ static int sr_class1p5_cdeinit(struct voltagedomain *voltdm,
 {
 	struct sr_class1p5_work_data *work_data;
 
+	if (IS_ERR_OR_NULL(voltdm) || IS_ERR_OR_NULL(class_priv_data)) {
+		pr_err("%s: bad parameters!\n", __func__);
+		return -EINVAL;
+	}
+
 	/* setup our work params */
 	work_data = get_sr1p5_work(voltdm);
 	if (IS_ERR_OR_NULL(work_data)) {
-- 
1.7.1

